test/micro/org/openjdk/bench/jdk/incubator/vector/Float16OperationsBenchmark.java

```diff
@@ -0,0 +1,296 @@
+        }
+        return res;
+    }
+
+    @Benchmark
+    public boolean isNaNBenchmark() {
+        boolean res = true;
+        for (int i = 0; i < vectorDim; i++) {
+            res &= isNaN(shortBitsToFloat16(vector1[i]));
+        }
+        return res;
+    }
+
+    @Benchmark
+    public void isNaNStoreBenchmark() {
+        for (int i = 0; i < vectorDim; i++) {
+            vectorPredicate[i] = Float16.isNaN(shortBitsToFloat16(vector1[i]));
+        }
+    }
+
+
+    @Benchmark
+    public void isNaNCMovBenchmark() {
+        for (int i = 0; i < vectorDim; i++) {
+            vectorRes[i] = Float16.isNaN(shortBitsToFloat16(vector1[i])) ? f16_one : f16_two;
+        }
+    }
+
+
+    @Benchmark
+    public void isInfiniteStoreBenchmark() {
+        for (int i = 0; i < vectorDim; i++) {
+            vectorPredicate[i] = Float16.isInfinite(shortBitsToFloat16(vector1[i]));
+        }
+    }
+
+
+    @Benchmark
+    public void isInfiniteCMovBenchmark() {
+        for (int i = 0; i < vectorDim; i++) {
+            vectorRes[i] = Float16.isInfinite(shortBitsToFloat16(vector1[i])) ? f16_one : f16_two;
+        }
+    }
+
+
+    @Benchmark
+    public void isFiniteStoreBenchmark() {
+        for (int i = 0; i < vectorDim; i++) {
+            vectorPredicate[i] = Float16.isFinite(shortBitsToFloat16(vector1[i]));
+        }
+    }
+
+
+    @Benchmark
+    public void isFiniteCMovBenchmark() {
+        for (int i = 0; i < vectorDim; i++) {
+            vectorRes[i] = Float16.isFinite(shortBitsToFloat16(vector1[i])) ? f16_one : f16_two;
+        }
+    }
+
+    @Benchmark
+    public void maxBenchmark() {
+        for (int i = 0; i < vectorDim; i++) {
+            vectorRes[i] = float16ToRawShortBits(max(shortBitsToFloat16(vector1[i]), shortBitsToFloat16(vector2[i])));
+        }
+    }
+
+    @Benchmark
+    public void minBenchmark() {
+        for (int i = 0; i < vectorDim; i++) {
+            vectorRes[i] = float16ToRawShortBits(min(shortBitsToFloat16(vector1[i]), shortBitsToFloat16(vector2[i])));
+        }
+    }
+
+    @Benchmark
+    public void sqrtBenchmark() {
+        for (int i = 0; i < vectorDim; i++) {
+            vectorRes[i] = float16ToRawShortBits(sqrt(shortBitsToFloat16(vector1[i])));
+        }
+    }
+
+    @Benchmark
+    public void negateBenchmark() {
+        for (int i = 0; i < vectorDim; i++) {
+            vectorRes[i] = float16ToRawShortBits(negate(shortBitsToFloat16(vector1[i])));
+        }
+    }
+
+    @Benchmark
+    public void absBenchmark() {
+        for (int i = 0; i < vectorDim; i++) {
+            vectorRes[i] = float16ToRawShortBits(abs(shortBitsToFloat16(vector1[i])));
+        }
+    }
+
+    @Benchmark
+    public void getExponentBenchmark() {
+        for (int i = 0; i < vectorDim; i++) {
+            rexp[i] = getExponent(shortBitsToFloat16(vector1[i]));
+        }
+    }
+
+    @Benchmark
+    public short cosineSimilarityDoubleRoundingFP16() {
```

Comment:
Hi @jatin-bhateja , I understand that the PR is in draft stage but just wanted to put out this comment so that these changes (if needed) are included in your next PS as well :) 
This test fails with an "invalid cast type" error in this function - https://github.com/openjdk/jdk/blob/dfa5620ff3d57b71ccaf09ca8e71fa85d93ceb00/src/hotspot/share/opto/castnode.cpp#L470. When it tries to compute `macRes` at line #238 the compiler generates a ConstraintCastNode and it fails to match the half_float type. We need to add isa_half_float() as another condition in this routine and define a new Cast node for half-float.

Reply:
ReplHF_short_reg is forward looking pattern for vector API use case, for auto-vectorization the one with constant and scalar are relevant and I have added test points for both the cases.